### DISCUSSION TITLE
bluez-alsa fullduplex sound

### DISCUSSION THREAD

**Discussion by leshiy1000** [Arkq/bluez-alsa#675] [General]
Hello, I connect bluetooth headphones in HSP mode, I get a lot of noise during playback and recording, is it possible to use full duplex in a2dp mode on one hci device for playback and recording?

**Comment by arkq** [comment]
Yes, it is possible to use full duplex A2DP as long as your devices support A2DP codec with full duplex. As for now, this project supports only FastStream codec which is capable of full duplex connection (aptx-LL is not supported yet).

**Comment by leshiy1000** [comment]
Hmm, I see, I need to figure out what my headphones support. Wasn't there a problem with the fact that when working through a2dp, the music is transmitted purely "aplay -D bluealsa:SRV=org.bluealsa,DEV=1C:91:9D:E0:0B:86,PROFILE=a2dp /tmp/f.wav ", but when working with HSP there is a lot of noise. example command "aplay -f S16_LE -D bluealsa:DEV=hci0,DEV=1C:91:9D:E0:0B:86,PROFILE=sco /tmp/f.wav" Moreover, there is noise both during recording and playback.

**Comment by leshiy1000** [comment]
Also, I still don’t quite understand how to work with HFP, I make a connection and receive a log:
[bluetooth]# pair 1C:91:9D:E0:0B:86
Attempting to pair with 1C:91:9D:E0:0B:86
[CHG] Device 1C:91:9D:E0:0B:86 Connected: yes
[CHG] Device 1C:91:9D:E0:0B:86 Bonded: yes
[CHG] Device 1C:91:9D:E0:0B:86 Modalias: bluetooth:v005Dp223Bd0100
[CHG] Device 1C:91:9D:E0:0B:86 UUIDs: 00001101-0000-1000-8000-00805f9b34fb
[CHG] Device 1C:91:9D:E0:0B:86 UUIDs: 00001108-0000-1000-8000-00805f9b34fb
[CHG] Device 1C:91:9D:E0:0B:86 UUIDs: 0000110b-0000-1000-8000-00805f9b34fb
[CHG] Device 1C:91:9D:E0:0B:86 UUIDs: 0000110c-0000-1000-8000-00805f9b34fb
[CHG] Device 1C:91:9D:E0:0B:86 UUIDs: 0000110e-0000-1000-8000-00805f9b34fb
[CHG] Device 1C:91:9D:E0:0B:86 UUIDs: 0000111e-0000-1000-8000-00805f9b34fb
[CHG] Device 1C:91:9D:E0:0B:86 UUIDs: 00001200-0000-1000-8000-00805f9b34fb
[CHG] Device 1C:91:9D:E0:0B:86 ServicesResolved: yes
[CHG] Device 1C:91:9D:E0:0B:86 Paired: yes
Pairing successful
[Mi True Wireless EBs Basic S]# trust 1C:91:9D:E0:0B:86
[CHG] Device 1C:91:9D:E0:0B:86 Trusted: yes
Changing 1C:91:9D:E0:0B:86 trust succeeded
[Mi True Wireless EBs Basic S]# connect 1C:91:9D:E0:0B:86
Attempting to connect to 1C:91:9D:E0:0B:86
Connection successful
[Mi True Wireless EBs Basic S]# bluealsa: W: Unsupported AT message: SET: command:+CCWA, value:1
bluealsa: W: Unsupported AT message: SET: command:+CLIP, value:1
bluealsa: W: Unsupported AT message: CMD: command:+CLCC, value:(null)


Then I want to play music but I freeze, maybe I need to run some other AT commands

**Comment by leshiy1000** [comment]
I assembled the bluez-alsa utility with debugging and noticed that there is a problem "bluealsa: 185.151596789: [645] E: ba-rfcomm.c:1540: AT handler IO error: Resource temporarily unavailable" this is when executing the script "hfp-ag-start 1C: 91:9D:E0:0B:86" from the thread https://github.com/Arkq/bluez-alsa/wiki/Using-BlueALSA-with-HFP-and-HSP-Devices

**Comment by leshiy1000** [comment]
I was able to implement pulseaudio into the project, the problem was solved.
